ngraph/core/src/pre_post_process/pre_post_process.cpp

```diff
@@ -0,0 +1,268 @@
+            auto channels_idx = layouts::channels(state.m_layout);
+            std::vector<std::size_t> v(state.m_layout.size(), 1);
+            v[channels_idx] = m_values.size();
+            shape = ngraph::Shape(v);
+        }
+        auto constant = ngraph::op::v0::Constant::create(ngraph::element::f32, shape, m_values);
+        constant->set_friendly_name(node->get_friendly_name() + "/scale/Divide_Factor");
+
+        auto new_op = std::make_shared<ngraph::op::v1::Divide>(node, constant);
+        new_op->set_friendly_name(node->get_friendly_name() + "/scale/Divide");
+        return new_op;
+    }
+};
+
+class PrePostProcessorInternalData {
+public:
+    PrePostProcessorInternalData(const std::shared_ptr<Function>& function) : m_function(function) {}
+    std::shared_ptr<Function> m_function = nullptr;
+    std::vector<InContextImpl> in_contexts;
+    InContextImpl& last_in() {
+        return in_contexts.back();
+    }
+};
+
+PrePostProcessorBase::PrePostProcessorBase(std::unique_ptr<PrePostProcessorInternalData>&& impl)
+    : m_impl(std::move(impl)) {}
+
+PrePostProcessorBase::~PrePostProcessorBase() = default;
+PrePostProcessorBase::PrePostProcessorBase(PrePostProcessorBase&&) = default;
+
+InContext PrePostProcessorBase::in() {
+    m_impl->in_contexts.push_back(InContextImpl());
+    return InContext(std::move(m_impl));
+}
+
+InContext PrePostProcessorBase::in(int port_index) {
+    m_impl->in_contexts.push_back(InContextImpl(port_index));
+    return InContext(std::move(m_impl));
+}
+
+InContext PrePostProcessorBase::in(std::string tensor_name) {
+    m_impl->in_contexts.push_back(InContextImpl(std::move(tensor_name)));
+    return InContext(std::move(m_impl));
+}
+
+OutProcessor PrePostProcessorBase::out() {
+    return OutProcessor(std::move(m_impl));
+}
+
+std::shared_ptr<Function> PrePostProcessorBase::build() {
+    // TODO: maybe wrap this logic with 'FunctionPass'???
+    for (auto in_context : m_impl->in_contexts) {
+        if (in_context.is_default()) {
+            auto param = m_impl->m_function->get_parameters().front();
+            auto consumers = param->output(0).get_target_inputs();
+            // TODO: use back propagation to identify preprocessing input shape/type/layout
+            auto new_param_type = param->get_element_type();
+            auto new_param_shape = param->get_partial_shape();
+            auto new_param_layout = PartialLayout();
+            if (in_context.m_tensor_data.m_type != ov::element::dynamic) {
+                new_param_type = in_context.m_tensor_data.m_type;
+            } else {
+                // Look for pre_process data
+                for (auto action : in_context.m_preprocess_data.m_actions) {
+                    if (std::dynamic_pointer_cast<PreProcessConvertElementType>(action)) {
+                        throw ngraph::ngraph_error("Can't convert element type from unknown type");
```

Comment:
here we have an assumption that no other preprocessing operations convert element types. It's true for now I believe, but in general back propagation can help with it

Reply:
Probably we should not raise an exception in this case and just leave 'user_tensor_element_type' as 'dynamic'. But can you please provide an example of back propagation?
Suppose we have function with 'i8' input parameter
`PrePostProcessor(f).in().preprocess().convert_element_type(f32).<doOther>().build()' - how we can guess source tensor element type using back propagation?